presto-hive/src/main/java/com/facebook/presto/hive/HiveSplitSource.java

```diff
@@ -104,7 +121,66 @@ void fail(Throwable e)
             return failedFuture(throwable.get());
         }
 
-        return future;
+        return future.thenApply(this::dynamicallyFilterSplits);
+    }
+
+    private List<ConnectorSplit> dynamicallyFilterSplits(List<ConnectorSplit> splits)
```

Comment:
I would apply dynamic filters lazily, so we do not increase the latency of short running queries.